### PR TITLE
schema: allow label and labelAbbr within harm tags

### LINKED ISSUE #1062: Allow label and labelAbbr as child elements of harm
It would be useful to have `harm/label` and `harm/labelAbbr` that function in a similar manner to  `verse/label` and `verse/labelAbbr`.

Here is an example of `verse/label` and `verse/labelAbbr` to display verse numbers for lyrics:

Screen Shot 2022-12-21 at 10 52 20 AM

<details>
<summary> Click to view MEI data for above example </summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2022-12-21T10:54:16" version="3.14.0-dev-8bfd8f7">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="mcnn14m">
    <score xml:id="s1by46sh">
     <scoreDef xml:id="sotzy5w" midi.bpm="400.000000">
      <staffGrp xml:id="s1unwmp2">
       <staffDef xml:id="staffdef-L1F1" n="1" lines="5">
        <clef xml:id="c1agi7k3" shape="G" line="2" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" n="1">
       <staff xml:id="staff-L1F1" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L4F1" dur="1" oct="4" pname="c" accid.ges="n">
          <verse xml:id="verse-L4F2" n="1">
           <label xml:id="l1hloihw">1.</label>
           <syl xml:id="syl-L4F2"> A</syl>
          </verse>
          <verse xml:id="verse-L4F3" n="2">
           <label xml:id="l14ayebl">2.</label>
           <labelAbbr xml:id="lkbfpb5">2.</labelAbbr>
           <syl xml:id="syl-L4F3"> B</syl>
          </verse>
         </note>
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L5">
       <staff xml:id="staff-L5F1N1" n="1">
        <layer xml:id="layer-L5F1N1" n="1">
         <note xml:id="note-L6F1" dur="1" oct="4" pname="c" accid.ges="n">
          <verse xml:id="verse-L6F2" n="1">
           <syl xml:id="syl-L6F2"> A</syl>
          </verse>
          <verse xml:id="verse-L6F3" n="2">
           <syl xml:id="syl-L6F3"> B</syl>
          </verse>
         </note>
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L7">
       <staff xml:id="staff-L7F1N1" n="1">
        <layer xml:id="layer-L7F1N1" n="1">
         <note xml:id="note-L8F1" dur="1" oct="4" pname="c" accid.ges="n">
          <verse xml:id="verse-L8F2" n="1">
           <syl xml:id="syl-L8F2"> A</syl>
          </verse>
          <verse xml:id="verse-L8F3" n="2">
           <syl xml:id="syl-L8F3"> B</syl>
          </verse>
         </note>
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L9">
       <staff xml:id="staff-L9F1N1" n="1">
        <layer xml:id="layer-L9F1N1" n="1">
         <note xml:id="note-L10F1" dur="1" oct="4" pname="c" accid.ges="n">
          <verse xml:id="verse-L10F2" n="1">
           <syl xml:id="syl-L10F2"> A</syl>
          </verse>
          <verse xml:id="verse-L10F3" n="2">
           <syl xml:id="syl-L10F3"> B</syl>
          </verse>
         </note>
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L11">
       <staff xml:id="staff-L11F1N1" n="1">
        <layer xml:id="layer-L11F1N1" n="1">
         <note xml:id="note-L12F1" dur="1" oct="4" pname="c" accid.ges="n">
          <verse xml:id="verse-L12F2" n="1">
           <syl xml:id="syl-L12F2"> A</syl>
          </verse>
          <verse xml:id="verse-L12F3" n="2">
           <syl xml:id="syl-L12F3"> B</syl>
          </verse>
         </note>
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L13">
       <staff xml:id="staff-L13F1N1" n="1">
        <layer xml:id="layer-L13F1N1" n="1">
         <note xml:id="note-L14F1" dur="1" oct="4" pname="c" accid.ges="n">
          <verse xml:id="verse-L14F2" n="1">
           <syl xml:id="syl-L14F2"> A</syl>
          </verse>
          <verse xml:id="verse-L14F3" n="2">
           <syl xml:id="syl-L14F3"> B</syl>
          </verse>
         </note>
        </layer>
       </staff>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

The first verse is given the label, and the second verse is given both a label and a labelAbbr:

```xml
   <note xml:id="note-L4F1" dur="1" oct="4" pname="c" accid.ges="n">
          <verse xml:id="verse-L4F2" n="1">
              <label xml:id="l1hloihw">1.</label>
              <syl xml:id="syl-L4F2"> A</syl>
          </verse>
          <verse xml:id="verse-L4F3" n="2">
              <label xml:id="l14ayebl">2.</label>
              <labelAbbr xml:id="lkbfpb5">2.</labelAbbr>
              <syl xml:id="syl-L4F3"> B</syl>
          </verse>
    </note>
```

The `label` entry causes the `syl` to be prefixed with the given text.   The label text is right-justified to the `syl` and will only be displayed at the current note.   `labelAbbr` is text that is prefixed to the starting `syl` after each system break.  Some editions number only the start of each verse, while others will repeat the verse number on every line (particularly when there are many verses and you might mix them up otherwise).

-----------------------------

The proposal for this issue is to make `harm/label` and `harm/labelAbbr` function similarly to that of `verse`.   Currently `harm` does not have `label` or `labelAbbr` allowed as children elements.  Their functions would be identical to their use in `verse`.  In harmonic analyses, such labels are used to indicate the active key for the harmonic data.

Here is an example from the discussion page https://github.com/rism-digital/verovio/discussions/3166

IMG_4081

The `A:` and `E:`  text:

Screen Shot 2022-12-21 at 11 20 24 AM

are `harm/label` elements.   The first line harmonic data (scale degrees) are with respect to A major, while the second row is with respect to E major.

### PR DESCRIPTION
**Pull request by WolfgangDrescher**
This is my first contribution to the music-encoding repo. I don't know the MEI schema well yet so any help and feedback is appreciated.

This PR allows `<label>` and `<labelAbbr>` whting `<harm>` tags similar to the the same functionality for `<verse>` . @craigsapp mentioned that this should be easy to implement so I hope that my changes are sufficient.

Closes https://github.com/music-encoding/music-encoding/issues/1062
Refs https://github.com/rism-digital/verovio/discussions/3166#discussioncomment-4452490

### CODE REVIEW

**Comment by ahankinson** [comment]
Looking at the discussion in #1062 I can see the motivation for this.

However, there is an important difference between `<verse>` and `<harm>`, in that `<harm>` allows textual content, while `<verse>` relies on markup as its content.

So you can do this:

`<harm>C</harm>`

But not this:

`<verse>Foo</verse>`

So it's not clear to me how `<label>` and `<labelAbbr>` would interact with textual content in this context. What is the expected behaviour with something like:

```
<harm>
    <label>My label</label>
    C
</harm>
```

@craigsapp since you suggested it, could you explain what you are thinking about this?

(PS: Thanks for the PR @WolfgangDrescher ! Much appreciated.)

**Review by ahankinson — COMMENTED** [review]
Would like to see a response to my comment about textual content and label interactions.

**Comment by craigsapp** [comment]
Oops, the post from 4 days ago quickly got lost in my inbox.

Here is an example of `<label>` being used in `<harm>`:

Screenshot 2023-02-14 at 7 36 29 AM

From https://musictheory.pugetsound.edu/mt21c/ModulationsWithDiatonicPivotChords.html (which has additional examples).

The circled text are `<labels>` that indicate the musical key in which the functional harmony symbols should be interpreted. 

For example, the `I` in the pickup measure means G major because of the `G:` label.  At the modulation point, the chord symbol `I` on the top line is still in G major, while the second line starts a new key as indicated by the `D:` label, which interprets the same chord as `IV`.

In other words, the `G:` text is not `<harm>` data itself but is describing `<harm>` data.  So the semantic analogy to verse numbers is similar, where the purpose of the number is to describe which verse the lyric line belongs to, but the numbers themselves are not lyrical text to be sung.

Graphical behavior of the harm labels are also similar to verse labels:  Note that the `G:` label is not aligned with a note above it but rather is attached to the `I` text instead.

**Comment by craigsapp** [comment]
This does look strange:

```xml
<harm>
    <label>My label</label>
    C
</harm>
```

-----------------

The best alternative that I can think of would be to use `<dir>`:

The `<dir>` could have a `@startid` that points to the `<harm>` content that it prefixes:

```xml
<harm xml:id="chord">I</harm>
<dir startid="#chord">G:</dir>
```

When a `<dir>` is attached to a harm in this way, then the renderer should infer that it is to be placed in front of the `<harm>` rather than above or below it.

Currently in verovio, such a `<dir>` text is not rendered when attached to a `<harm>` via `@startid`, but it could be enhanced to make the `<dir>` function in a manner similar to a verse label.

<details>
<summary> Click to view sample encoding of dir being used as harm prefix text </summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2023-02-14T08:27:28" version="3.15.0-dev-2f1ce00-dirty">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="m119g8nq">
    <score xml:id="sdwcinj">
     <scoreDef xml:id="s1ddmktx">
      <staffGrp xml:id="s1coyiuz">
       <staffDef xml:id="staffdef-L1F1" n="1" lines="5">
        <clef xml:id="c151q1x4" shape="G" line="2" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1">
       <staff xml:id="staff-L1F1" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L4F1" dur="2" oct="5" pname="c" accid.ges="n" />
         <note xml:id="note-L5F1" dur="2" oct="4" pname="g" accid.ges="n" />
        </layer>
       </staff>
       <dir xml:id="dir-L3F1" place="below" staff="1" startid="harm=L4F2">
        <rend xml:id="r124ebao" fontstyle="normal">C:</rend>
       </dir>
       <harm xml:id="harm-L4F2" place="below" staff="1" tstamp="1.000000" n="2">I</harm>
       <harm xml:id="harm-L5F2" place="below" staff="1" tstamp="3.000000" n="2">V</harm>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

Dir element from the example:

```xml
       <dir xml:id="dir-L3F1" place="below" staff="1" startid="harm=L4F2">
```

Adding `@place` would be questionable (it should be taken from the harm it is attached to; likewise for staff, and the vertical position should also be taken from the harm (`@n` attribute).

**Comment by kepper** [comment]
We had a looong discussion on this at today's ODD meeting. I'm trying to summarize that. First, there is definitive room for improvement, so thanks for bringing this up ;-). The challenge is that `<harm>` isn't particularly structured / machine-readable anyway (that was the reason to bring up the `<fb>` etc. elements a couple of years ago). We therefore tried to split this up into two separates components to discuss. One is the horizontal alignment of multiple harms, the other one is the labelling of such groups. 

For the alignment problem, we suggest to just use the `@n` attribute, in a similar way to how it is used on `<verse>`. More explicit methods would include `@plist`, `@next`/ `@prev`, or some such.

The other questions seems a bit trickier. Adding `<label>` doesn't really seem to solve more problems than it creates. basically, one might argue that the circled _A_ label that gives the key for the following interpretations is a harmonic interpretation in itself. The circle around this could (assumably…) be captured using `<rend>` inside harm. 

These proposals doen't really improve the situation – which seems to be a much larger undertaking than what is currently discussed here – but may perhaps give an indication on how to deal with the situation with the existing concepts. Of course, this is just a proposal…

**Comment by craigsapp** [comment]
I would suggest the solution in my previous post, which solves the problem by using `<dir>` instead of creating a harm label:

```xml
 <harm xml:id="chord" n="2" startid="somenote">I</harm>
 <dir startid="#chord">G:</dir>
```

Typcially `<dir>` is attached to a note/rest.  By attaching it to a `<harm>` it is now a `<dir>` for the `<harm>`.   This can be inferred have the `<dir>` act as a `<label>` for the `<harm>`.  In this way, MEI does not need to be altered, and it is up to the rendering software (verovio) to handle it properly.


> The other questions seems a bit trickier. Adding `<label>` doesn't really seem to solve more problems than it creates. basically, one might argue that the circled A label that gives the key for the following interpretations is a harmonic interpretation in itself. The circle around this could (assumably…) be captured using `<rend>` inside harm.

There are semantic problems with this solution:  The key label is like an adjective, and `<harm>` are considered like nouns.   This is a similar situation for `<verse>` and `<label>`.    The key label should be applied to all subsequent `<harm>` at the same `@n` until another is encountered.  And there is the visual aspect: The key label comes before the note/timestamp on which the `<harm>` is attached, so it is visually distinct from the main `<harm>` content.

**Comment by pe-ro** [comment]
I recommend adding `@vgrp` to `<harm>` to capture horizontal alignment rather than relying on other attributes, such as `@n`.  It may be useful to also add `@vgrp` to any other elements that currently have `@vo`, by making their att.[elementName].vis class a member of att.verticalGroup.

As for labeling harmonic indications, while I agree that a mixed content model that includes `<label>` but no other elements is a little weird; however, it's still just a mixed model, and those occur throughout MEI.  Processing of mixed content models always depends on the order of elements and text inside, so the label can be displayed anywhere within the `<harm>` element.  Though almost all of the time the label should precede the other content, there's no easy way to force it to be encoded that way while retaining a mixed model.  Verovio and other processors are free to implement rules that always display label elements first, though it's probably not necessary as encoders will likely put the `<label>` at the beginning due to the visual order of the contents of `<harm>`.

Note that I said "display the label elements first" above as there is also no way to limit the number of `<label>` elements while retaining a mixed model.  Again, I don't think this will present a significant problem.

**Comment by craigsapp** [comment]
Adding formalized structure for `<harm>` will be complicated since there are many systems of describing harmonic information, both visually and syntactically.   If `<label>` were present inside of `<harm>`, then perhaps this is the best solution:

```xml
<harm>
    <label>
    <harm>text content of harm</harm>
</harm
```

There would be no CDATA allowed in the first harm but only regular elements when a label is present.  The free-form harm data would be stored in the child harm.   Or the free-form content of harm could be required to be stored in `<rend>` in such cases:

```xml
<harm>
    <label>
    <rend>text content of harm</rend>
</harm
```

But I think the linked `<dir>` solution is sufficient and should work well.

**Comment by pe-ro** [comment]
I think the linked solution is a little cumbersome, but allowing `<label>` within `<harm>` doesn't preclude it.

**Comment by craigsapp** [comment]
> I recommend adding `@vgrp` to `<harm>` to capture horizontal alignment rather than relying on other attributes, such as `@n`. It may be useful to also add `@vgrp` to any other elements that currently have `@vo`, by making their `att.[elementName].vis` class a member of `att.verticalGroup`.

Verovio currently uses `@n` for vertical placement of `<harm>`.  This is working well, and preferably it would not be changed (but adding `@vgrp` as an alternate system is not a problem, and when both are given, then one or the other can be given priority -- probably `@vgrp` which would control the visual placement of the harm, while `@n` would indicate the logical grouping of successive harm).

**Comment by lpugin** [comment]
I am wondering if this is not something that could go in `<scoreDef>`. Something like:
```xml
<scoreDef>
   <staffGrp>
      <staffDef n="1">
         <harmDef n="1">
            <label>G major:</label>
            <labelAbbr>G:</labelAbbr>            
         </harmDef>
         <harmDef n="1">
            <label>D major:</label>
            <labelAbbr>D:</labelAbbr>            
         </harmDef>
      </staffDef>
   </staffGrp>
</scoreDef>
```
Certainly more complicated and more refactoring (sorry about it), but maybe cleaner?

We could do the same for verse:
```xml
<scoreDef>
   <staffGrp>
      <staffDef n="1">
         <verseDef n="1">
            <label>Verse I</label>
            <labelAbbr>I</labelAbbr>
         </verseDef>
      </staffDef>
   </staffGrp>
</scoreDef>
```

As far as I know, using `@n` for the alignment of `harm` is currently not an issue, so I am not sure we really need `@vgrp`. Furthermore, the solution I am proposing, or something along  the same line, would probably imply that `@n` is used for the alignment.

**Comment by craigsapp** [comment]
I am presuming `@n="2"` for the second harm label:

```xml
<scoreDef>
   <staffGrp>
      <staffDef n="1">
         <harmDef n="1">
            <label>G major:</label>
            <labelAbbr>G:</labelAbbr>            
         </harmDef>
         <harmDef n="2">
            <label>D major:</label>
            <labelAbbr>D:</labelAbbr>            
         </harmDef>
      </staffDef>
   </staffGrp>
</scoreDef>
```

This might work, although it is not generalizable to all harmony analysis cases.   For modulations there can be a region of common chord labels that are shown on two lines, but also this can be shortened to a single chord sequence with the modulation occurring in the same `harm@n` stream:

Screenshot 2023-03-08 at 6 28 22 AM

From: https://en.wikipedia.org/wiki/Modulation_(music)#/media/File:Chromatic_modulation_in_Bach_BWV_300,_m._5-6.png
https://en.wikipedia.org/wiki/Modulation_(music)

In this particular case there is a harmony label that changes within the middle of a measure (which cannot be encoded via scoreDef/staffDef as far as I know).

**Comment by lpugin** [comment]
You are right that this is not generalizable. In the end it would probably be best to use positioning and other existing attributes such as `@type`. You could do something like this:

image

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
   <meiHead>
      <fileDesc>
         <titleStmt>
            <title></title>
         </titleStmt>
         <pubStmt>
            <date>2017-04-25</date>
         </pubStmt>
         <seriesStmt>
            <title>Verovio test suite</title>
         </seriesStmt>
      </fileDesc>
      <encodingDesc>
         <appInfo>
            <application version="2.0.0" label="2">
               <name>Verovio</name>
            </application>
         </appInfo>
      </encodingDesc>
   </meiHead>
   <music>
      <body>
         <mdiv>
            <score>
               <scoreDef>
                  <staffGrp bar.thru="true">
                     <staffDef n="1" lines="5" clef.shape="G" clef.line="2" />
                     <staffDef n="2" lines="5" clef.shape="F" clef.line="4" />
                  </staffGrp>
               </scoreDef>
               <section>
                  <measure n="1">
                     <staff n="1">
                        <layer n="1">
                           <note dur="2" oct="4" pname="b" />
                           <note dur="2" oct="4" pname="b" />
                        </layer>
                        <layer n="2">
                           <note dur="1" oct="4" pname="g" />
                        </layer>
                     </staff>
                     <staff n="2">
                        <layer n="1">
                           <note dur="2" oct="3" pname="e" />
                           <note dur="2" oct="3" pname="e" />
                        </layer>
                     </staff>
                     <harm place="below" staff="2" tstamp="0.500000" n="1" type="label">A:</harm>
                     <harm place="below" staff="2" tstamp="1.00000" n="1">i</harm>
                     <harm place="below" staff="2" tstamp="2.50000" n="1" type="label">A:</harm>
                     <harm place="below" staff="2" tstamp="3.00000" n="1">i</harm>
                  </measure>
               </section>
            </score>
         </mdiv>
      </body>
   </music>
</mei>
```

**Comment by craigsapp** [comment]
Yes, in the discussion for issue https://github.com/rism-digital/verovio/issues/3307 I am using `@type="key-label"` to allow semantic differentiation between key labels and chord labels in `<harm>` data.


```xml
<harm xml:id="h4ee7f7" type="key-label" place="below" staff="2" tstamp="1.000000" n="1">
    <rend xml:id="rxtwmhm" halign="right">
    <rend xml:id="rg7cq5v" fontname="sans-serif">A:</rend> </rend>
</harm>
```

I still think that it is preferable to allow `@halign="right"` for the key-label `<harm>` rather than to tweak its `@tstamp` value.   In particularly because the tweaked timestamp is also dependent on the previous harm label's timestamp, and if the key-label timestamp is too close to the next harm element, I am seeing spacing problems in the verovio rendering.  In addition, the rhythm in other parts on the system can alter the position of the key-label if adjusting the `@tstamp` of the key-label.

The space between the colon and the chord should be the same for all modulations, so clearly the key-label needs a `halign="right"` with a shared timestamp with the chord (given with implicit `haligh="left"`).    Here is an illustration of the problem:

Screenshot 2023-03-08 at 8 18 59 AM

<details>
<summary> Click to view MEI data for above example </summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2023-03-08T08:17:57" version="3.16.0-dev-9aae300">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="m1mmdrv4">
    <score xml:id="s1vpwysg">
     <scoreDef xml:id="sia65rp">
      <staffGrp xml:id="s1fmg8dd" bar.thru="true" symbol="brace">
       <staffDef xml:id="staffdef-L1F3" n="1" lines="5">
        <clef xml:id="cb8uo9b" shape="G" line="2" />
       </staffDef>
       <staffDef xml:id="staffdef-L1F2" n="2" lines="5">
        <clef xml:id="c1trn7un" shape="F" line="4" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1">
       <staff xml:id="staff-L1F3" n="1">
        <layer xml:id="layer-L1F3N1" n="1">
         <note xml:id="note-L4F3" dur="4" oct="4" pname="b" accid.ges="n" />
         <beam xml:id="beam-L5F3-L9F3">
          <note xml:id="note-L5F3" dur="16" oct="5" pname="c" accid.ges="n" />
          <note xml:id="note-L6F3" dur="16" oct="5" pname="d" accid.ges="n" />
          <note xml:id="note-L7F3" dur="16" oct="5" pname="c" accid.ges="n" />
          <note xml:id="note-L8F3" dur="32" oct="4" pname="b" accid.ges="n" />
          <note xml:id="note-L9F3" dur="32" oct="4" pname="a" accid.ges="n" />
         </beam>
         <note xml:id="note-L11F3" dur="2" oct="4" pname="b" accid.ges="n" />
        </layer>
        <layer xml:id="layer-L1F4N2" n="2">
         <note xml:id="note-L4F4" dur="1" oct="4" pname="g" accid.ges="n" />
        </layer>
       </staff>
       <staff xml:id="staff-L1F2" n="2">
        <layer xml:id="layer-L1F2N1" n="1">
         <note xml:id="note-L4F2" dur="2" oct="3" pname="e" accid.ges="n" />
         <note xml:id="note-L11F2" dur="2" oct="3" pname="e" accid.ges="n" />
        </layer>
       </staff>
       <harm xml:id="harm-L7F1" place="below" staff="2" tstamp="0.5" n="1">
        <rend xml:id="rgdw8wr">A:</rend>
       </harm>
       <harm xml:id="harm-L4F1" place="below" staff="2" tstamp="1.000000" n="1">
        <rend xml:id="rpieg1m">i</rend>
       </harm>
       <harm xml:id="harm-L7F1" place="below" staff="2" tstamp="2.500000" n="1">
        <rend xml:id="rgdw8wr">A:</rend>
       </harm>
       <harm xml:id="harm-L11F1" place="below" staff="2" tstamp="3.000000" n="1">
        <rend xml:id="r1i6gh1b">i</rend>
       </harm>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

Notice that the distance between the `A:` and the `i` are different.   There can be other corner cases that can cause problems, such as clef changes between the key-label and chord harms when they use different tstamps.

**Comment by pe-ro** [comment]
A minor comment:  The two instances of `<rend xml:id="rgdw8wr">A:</rend>` in the example above have the same `@xml:id`.  That will keep the file from validating.

**Comment by craigsapp** [comment]
Thanks!  In this case I was copy and pasting the harm at `@tstamp="0.5"`, but if generated automatically I would hope that the `xml:id`s would be different.

**Comment by ahankinson** [comment]
05.2023: After extensive discussion of this issue at the developers conference no consensus could be reached on this PR. (@ahankinson, @pe-ro, @jcdevaney participating)

The only conclusion that could be reached is that the "G: " and "D: ", given in https://github.com/music-encoding/music-encoding/pull/1070#issuecomment-1429984480 are also themselves harmonic labels, so wrapping them in `<harm>` tags is not wrong.

From there, attributes such as `@next` or `@follows` could be used to link harmonic labels together. `@plist` can also be used to point to a list of other harmonic labels that are part of a group. The positioning attributes can be used to visually place the harmonic label. (This follows suggestions made earlier in this discussion from the ODD meetings.)

So the conclusion was that this sort of encoding could be achieved with the existing markup capabilities.

In the end we decided to leave this PR open, report on our discussions, and request further feedback, but put this in the MEI 6 milestone.

**Comment by craigsapp** [comment]
Here is the current solution I am using:

Screenshot 2023-05-29 at 10 46 45 PM


<details>
<summary> Click to view MEI data for above example </summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2023-05-29T22:44:06" version="3.16.0-dev-b62cd9f-dirty">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="m1687xhc">
    <score xml:id="swc03tu">
     <scoreDef xml:id="s7q99va">
      <staffGrp xml:id="so18ux">
       <staffDef xml:id="staffdef-L1F1" n="1" lines="5">
        <clef xml:id="cidki1p" shape="G" line="2" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1">
       <staff xml:id="staff-L1F1" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L3F1" dur="4" oct="4" pname="c" accid.ges="n" />
         <note xml:id="note-L4F1" dur="4" oct="4" pname="e" accid="f" />
         <note xml:id="note-L5F1" dur="4" oct="4" pname="g" accid.ges="n" />
        </layer>
       </staff>
       <harm xml:id="harm-L3F2" place="below" staff="1" tstamp="1.000000" n="2">
        <rend xml:id="r1fq3u8h">i</rend>
       </harm>
       <harm xml:id="htw6pxb" type="key-label" place="below" staff="1" tstamp="1.000000" n="2">
        <rend xml:id="rlsmmp4" halign="right">
         <rend xml:id="rg44est" color="darkred" fontname="sans-serif">C:</rend> </rend>
       </harm>
      </measure>
      <scoreDef xml:id="s1lv2bwl">
       <keySig xml:id="keysig-L7F1" sig="2s" />
      </scoreDef>
      <measure xml:id="measure-L6">
       <staff xml:id="staff-L6F1N1" n="1">
        <layer xml:id="layer-L6F1N1" n="1">
         <note xml:id="note-L9F1" dur="2" oct="4" pname="a" accid.ges="n" />
         <note xml:id="note-L10F1" dur="2" oct="5" pname="c" accid.ges="s" />
        </layer>
       </staff>
       <harm xml:id="harm-L9F2" place="below" staff="1" tstamp="1.000000" n="2">
        <rend xml:id="r1v5oq3l">V</rend>
       </harm>
       <harm xml:id="hlfrbgf" type="key-label" place="below" staff="1" tstamp="1.000000" n="2">
        <rend xml:id="r2pkscf" halign="right">
         <rend xml:id="r4tgb69" color="darkred" fontname="sans-serif">D:</rend> </rend>
       </harm>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

This works reasonable well.  I place the two labels at the same timestamp, and add `rend@halign="right"` to the key label as well as `harm@type-"key-label"`.   (Currently a spacing bug in verovio (issue https://github.com/rism-digital/verovio/issues/3307) where it adds extra space after such a group.)  Maybe setting `harm@form="key-label" would automatically apply the style the text inside of it to `rend@halign="right"` would be a good thing to have.

**Comment by ahankinson** [comment]
Thanks @craigsapp -- would you say that this solution is good enough and that this change is no longer needed?

@WolfgangDrescher it would be great to hear from you too.

**Comment by craigsapp** [comment]
I would say that this solution is good one if @lpugin likes it.  

The main concern is about the sequence of the data that occurs at the same time (i.e., if there is a key label at the same time as the chord symbol, which one comes first.  I would say that it should be safe to infer that the chord label comes before the chord symbol, or in other words the chord symbol should be interpreted in the key label that occurs at the same time (rather than the previous key label). 

It might be good to add linking attributes to make this clear (since the chord symbol could appear in the MEI data before the key label), but I don't see this as necessary (it would be more convenient however for analysis): you would extract the `@tstamp` of every `<harm>` and then sort by the `harm@type` when elements are at the same time with the same `@n` when processing the data for harmonic analysis.

**Comment by ahankinson** [comment]
> It might be good to add linking attributes to make this clear (since the chord symbol could appear in the MEI data before the key label), but I don't see this as necessary (it would be more convenient however for analysis): you would extract the @tstamp of every <harm> and then sort by the harm@type when elements are at the same time with the same @n when processing the data for harmonic analysis.

This might be tricky if you have non-related harmonic analyses occurring at the same time... 

`harm@plist` is a useful arbitrary grouping mechanism, so you could link all "members" of a chord label using that. Failing that you can do a linked-list with `harm@next` and `harm@prev`, and the first one of this is the chord label. (The one with no `@prev` attribute) This would mean you could ignore the tstamp, or process stuff where the `harm` is linked to the note ID rather than a tstamp.

**Comment by lpugin** [comment]
The problem in Verovio has been [fixed](https://github.com/rism-digital/verovio/issues/3307), so it seems to me that the PR can be closed.

**Comment by lpugin** [comment]
Since there was no follow up I am closing it. Thanks for the contribution, though.
